core/src/package_store.rs

```diff
@@ -0,0 +1,559 @@
+    pub fn finalize(
+        &mut self,
+        #[cfg(feature = "native")] engine: &Engine,
+    ) -> Result<(), Vec<CoreError>> {
+        // First, get all successful fetches and add them to external_packages, propagating any
+        // error when creating the package itself
+        let (successes, failures) = self
+            .new_packages
+            .drain()
+            .map(|(k, v)| package_new!(v.as_slice(), engine).map(|p| (k, p)))
+            .fold((vec![], vec![]), |(mut s, mut f), res| {
+                match res {
+                    Ok(ok) => s.push(ok),
+                    Err(err) => f.push(err),
+                };
+                (s, f)
+            });
+
+        for (package_name, package) in successes {
+            self.external_packages.insert(package_name, package);
+        }
+
+        // Here we check if any packages failed to compile (not if they failed to resolve)
+        if !failures.is_empty() {
+            return Err(failures);
+        }
+
+        // Then, clear all failed packages and if any failed, return that error
+        // Note that "drain" here drains all errors, not just the one we (possibly) return
+        let failed = mem::take(&mut self.failed_packages);
+        if failed.is_empty() {
+            Ok(())
+        } else {
+            Err(failed)
+        }
+    }
+
+    /// This function loads the default packages to the Context. First, it loads all native
+    /// packages, retrieved from `std_packages::native_package_list()`, and then it loads all
+    /// standard packages by passing this Context to `std_packages::load_standard_packages()`. This
+    /// will be run when constructing the Context, and may only be run once.
+    pub(crate) fn load_default_packages(
+        &mut self,
+        #[cfg(feature = "native")] engine: &Engine,
+    ) -> Result<(), CoreError> {
+        for pkg in std_packages::native_package_list() {
+            self.load_native_package(Package::new_native(pkg)?)?
+        }
+        #[cfg(feature = "native")]
+        std_packages::load_standard_packages(self, engine)?;
+        #[cfg(not(feature = "native"))]
+        std_packages::load_standard_packages(self)?;
+        Ok(())
+    }
+
+    /// This function loads a package from the serialized format retrieved from Module::serialize.
+    /// This is only available on native when using the precompile_wasm feature.
+    #[cfg(all(feature = "native", feature = "precompile_wasm"))]
+    pub(crate) fn load_precompiled_standard_package(
+        &mut self,
+        wasm_source: &[u8],
+        engine: &Engine,
+    ) -> Result<(), CoreError> {
+        let pkg = Package::new_precompiled(wasm_source, engine)?;
+        self.insert_standard_package(pkg)
+    }
+
+    /// This is a helper function to load a package directly from its wasm source. It will be
+    /// compiled using `Package::new` to become a `Package` and then loaded using `load_package`
+    #[allow(dead_code)]
```

Comment:
This is needed since if we are using `bundle_std_packages` but not `precompile_wasm`, it is needed, and if we are using `precompile_wasm`, this isn't used; this avoids the compiler warning